apps/passport-server/src/database/queries/rateLimit.ts

```diff
@@ -1,40 +1,122 @@
 import { Pool } from "postgres-pool";
+import { RateLimitBucket } from "../models";
 import { sqlQuery } from "../sqlQuery";
 
 /**
- * Calls the `take_token` stored procedure to determine if a "token" is
- * available. If so, the action is permitted, otherwise it is denied due to
- * exceeding the rate limit.
+ * Determines if a rate-limited action should be allowed to proceed.
  *
- * See 58_rate_limit.sql
+ * The rate-limiting mechanism here is a "token bucket". The bucket contains a
+ * number of "tokens" (the `remaining` value in the rate_limit_buckets table).
+ * Clients attempt to consume tokens, and if a token is available then the
+ * action is permitted. If no token is available, it is denied.
+ *
+ * Tokens "refill", allowing the user to perform additional actions over time.
+ * For instance, if maxActions is 10, and timePeriod is 86400000 (60 minutes in
+ * ms), then tokens refill at a rate of one every 6 minutes.
+ *
+ * This allows clients to continue to make requests, but only at a limited
+ * rate. If the client waits for a whole hour, then 10 tokens will be due,
+ * fully replenishing the bucket.
+ *
+ * Token refills are not performed on a timer, but instead we calculate how
+ * many tokens are due to be refilled whenever we process a request, by
+ * calculating the difference between the current time and the last time we
+ * consumed a token, and adding the appropriate number of tokens.
  *
  * @param client
  * @param actionType The type of action being attempted
  * @param actionId The identifier of this action
+ * @param maxActions The maximum number of actions allowed in a time period
+ * @param startingActions The number of actions to start the bucket with
+ * @param timePeriod The time period in which maxActions are allowed
  * @returns Boolean indicating whether the action is permitted
```

Comment:
Doesn't actually return a boolean.  Actually returns a bucket, requiring the caller to check the `remaining` value themselves.

Reply:
Fixed!